### DISCUSSION TITLE
Fixing "high indexing load reference rate" issues

### DISCUSSION THREAD

**Discussion by nemanja-tosic** [ravendb/ravendb#16390] [Q&A]
Hello!

We are working on a project management tool which consists of documents and pages, where a document contains many (many) pages. A document can be accesses by users that were invited to them. We eagerly load all documents (and their constituent pages) and stream them back to the user. The streaming bit is important - we want to do a single query and stream the responses back.

The query we use goes something like:
```
from index DocumentsPagesAndUsers where partOf in ("userId")
```
or in other words what we're trying to do is return everything that a certain user needs in order to use the app. What a user has access to is subject to change over time. It's important to note that the user needs to fetch Document, Page and User entities in a single query.

In order to make this query work, as far as i know, DocumentsPagesAndUsers index needs to have a partOf property with the "userId" inside. This can either be achieved (as far as i know) by having the field partOf "generated" during indexing by indexing related data or having the partOf updated for all documents when needed.

Updating the documents is a no-go - a document can have 20 000 pages easily, and we'd like to avoid concurrency issues, if nothing else. A user might have hundreds of documents, so the issue snowballs.

Indexing related data is where we get into hot water... users that can access a Document are stored on the Document meaning that whenever a Document changes (because we've added/removed someone) 20 000 documents need to be updated and we get the "high indexing load reference rate" warning and the index takes a while to update (i believe upwards from 20 seconds for a relatively small result set).

Is there anything we're missing here, or maybe we modeled our database poorly for this use case?

To counter the high reference load, we've then went on to create a map/reduce index, in order to remove references altogether. However, reductions are problematic - Pages don't know who they belong to apart from a Document so we either need to create a document reference (exactly what we wanted to avoid) or to group by document id so that we can aggregate the partOf aspect. Grouping however produced one document so we cannot stream effectively anymore.

We then landed on the following:

```
# NOTE: only Documents and Pages have the "documentId" property

# Map documents and pages
from entity in DocumentsPagesAndUsers
where entity.documentId != null
select new {
    entityId = entity.Id,
    groupId = entity.documentId,
    partOf = entity.partOf,
    entities = new [] { entity.Id }
}

# Map everything else
from entity in DocumentsPagesAndUsers
where entity.documentId == null
select new {
    entityId = entity.Id,
    groupId = entity.Id,
    partOf = entity.partOf.Concat(new [] { entity.Id }),
    entities = new [] { entity.Id }
}

# Reduce
from result in results
group result by new { result.groupId, result.entityId } into g
let partOf = g.SelectMany(x => x.partOf).Distinct()
let entities = g.SelectMany(x => x.entities).Distinct()
# NOTE THE SECOND FROM
from entity in entities
select new {
    groupId = g.Key.groupId,
    entityId = g.Key.entityId,
    partOf = partOf,
    entities = new [] { entity }
}
```

If you are still here, thank you!

From the point of view and functionality and performance (almost!) this is exactly what we need! However, initial index generation takes minutes to complete for a small result set (~40k docs.)

So, the questions we have here are:
* is a from in a reducer a bad thing? It achieves our exact goals and acts as sort of a join
* if so what are the alternatives we have, as we feel we have exhausted the possibilities with the basic building blocks we have?
* if not - why does it take so long to build the index?

This turned out to be a longer post than i expected, but there was a lot of ground to cover. Happy to elaborate/summarize anything that is unclear. Thanks!

**Comment by ayende** [comment]
Yes, this isn't a good idea. You are _not_ reducing, to start with.
And you are forcing us to regenerate the data on each document update, which is not a good idea.
If I understand things, you have:

* Documents
* Pages

A page belongs to a document. Probably using something like `documentId`, I assume?
Not sure if they are in a single collection or separate. 

A user may be granted access to a particular page, or to a document. If they belong to a document, they have access to the all the associated pages. Is that correct?

How often do those permission change? How many unique users have access to a document?
Do you have roles as well?

Are you always operating within the scope of a single document?

**Reply by nemanja-tosic** [reply]
Correct, we are not reducing per se, we are using the reducer to enhance Page's partOf property. The initial idea was to reduce to a "lookup"-styled document, which then meant we couldn't stream as all results ended up in a single document. This is how we ended up with a from in a reducer - "how can we produce this lookup but also be able to stream results later".

Updates to the index work extremely fast from our experiments, the initial load is the problematic one if that gives any insight.

For completeness, we have:

* User
* Document
* Page

A Page belongs to a Document, and both Page and Document contain a `documentId` property. Currently everything is in the same collection, we were going for a generic loading mechanism by utilizing parent references (Page -> Document -> User). The generic mechanism means a recursive query until you hit a "null" though.

Users are granted access to Documents, if they have access to Document, they have access to all Pages, correct.

It is RBAC based system so we have roles, and the permissions currently are kept on a Document. The permissions change infrequently. Documents also change quite infrequently, so moving the permissions out doesn't necessarily yield much.

The number of users can be large, but back of the envelope answer would be i'd like to support 50 users.

Users have access to many Documents and would work on all of them.

**Reply by ayende** [reply]
When you stream the pages of a document, do you do that on a _single_ document? Or something like: "Give me all the pages of all the documents for this user" ?

**Reply by nemanja-tosic** [reply]
All the Pages of all the Documents for a User.

We have optimized the backend so that data doesn't touch the backend server - the server verifies if the user has access to what they are requesting and then streams the result back to the client.

**Reply by ayende** [reply]
Not sure that I'm following here. What do you mean, backend server here?

**Reply by nemanja-tosic** [reply]
What i meant was, we're trying to do as little as possible when fetching the result set, basically we take a "passthrough" approach. We have a Client (frontend), API (backend) and the DB Server. The backend acts as a pipeline basically - get a stream from raven, pipe it through to the client, no processing in between.

It's not necessarily important to the overarching conversation, more to say that we stream the entire result set and had success with the approach so far.

**Reply by mcrio** [reply]
> Updating the documents is a no-go - a document can have 20 000 pages easily, and we'd like to avoid concurrency issues, if nothing else. A user might have hundreds of documents, so the issue snowballs.

Maybe I did not understand the approach correctly, but do you think it's going to scale? Streaming all data for a user would be something like 20,000 pages * 100 documents = 2,000,000 records and more.. It surely depends on the infrastructure (intranet, internet), number of users, etc.. but still seems a bit unusual.

**Reply by mcrio** [reply]
> //Map documents and pages
> from entity in DocumentsPagesAndUsers
> where entity.documentId != null

> //Map everything else
> from entity in DocumentsPagesAndUsers
> where entity.documentId == null

Just to clarify, what would be "everything else" when the first comment says that pages and documents have `documentId != null` ?

I guess if you could make a simple test case with autogenerated dummy data, maybe that would help to understand the issue better.

**Reply by nemanja-tosic** [reply]
> Maybe I did not understand the approach correctly, but do you think it's going to scale? Streaming all data for a user would be something like 20,000 pages * 100 documents = 2,000,000 records and more.. It surely depends on the infrastructure (intranet, internet), number of users, etc.. but still seems a bit unusual.

It will not scale in the current form, however right now we have users that have a maximum of 30k documents to stream. This amounts to roughly < 10MB of data and is sufficiently fast. Moving forward i believe streaming will still work, however it will need to be tweaked.

In the current form we would like to preload our query model and we can say that the size of the set is max 30k documents, which should be well within reasonable limits.

Paginated load is not something we would like to do right now, especially seeing how the size of data is well within our average user bandwith capability.

Whether or not we stream, the issue we have with the index remains though.

**Reply by nemanja-tosic** [reply]
> Just to clarify, what would be "everything else" when the first comment says that pages and documents have documentId != null ?

The last entity in the example domain is a User - when fetching everything that a User should have access to, we return the User as well.

I can produce a test case to demonstrate the issue i suppose, sure.

**Reply by nemanja-tosic** [reply]
Here is a representative example of what we want: [Dump of partof 2023-05-03 18-52.ravendbdump.zip](https://github.com/ravendb/ravendb/files/11385021/Dump.of.partof.2023-05-03.18-52.ravendbdump.zip)

We are able to load everything we need by doing:
```
from index "QueryModel/ByPartOf" where partOf in ('users/1')
```

Testing locally yields better results to what we have in prod - for this result set i get 2s locally compared to 7s needed for 13k documents in prod. The stage that is noticeably longer in "index performance" is the LoadDocument stage, so the example is only representative of the general issue of too many references, it is a lot slower with real-world data.

The warning we're trying to resolve, and in turn hopefully the performance issues is: 

![image](https://user-images.githubusercontent.com/10130001/235986147-7e3c2af0-b7dc-47a3-8375-c5040b367eb4.png)

**Reply by ayende** [reply]
Thanks, that is really helpful. 
However, your model & the data do not match.

You have a *flat* model for the data in what you sent.

*All* `pages` documents have a `partOf` = 'documents/1'

Is that typical of your model? Or do you have a true nested structure?
Book > Part > Chapter > Page - etc?

The core issue is that you are trying to query from the user to all the association. And that means that you need to follow the link from the user to the relevant documents. That would also be a better match to how you *usually* work.

Streaming tens of thousands or more documents to the user is *expensive*. If nothing else, RavenDB would need to read a lot of individual documents. The issue is that separating this to a two step process:

```
from QueryModel where type = 'Document' and partOf = $user
select documentId
```

And

```
from QueryModel where partOf in ($documentIds)
```

Would be far easier and cheaper

**Reply by nemanja-tosic** [reply]
The data i've sent is a representative example of what we have, it is not complete.

The goal is to have a generic way to preload portions of the query model with a single query - i.e. get all the documents that a user needs to use the app without any further loads. `Document` is not the only top-level entity that a user needs to fetch on load.

The solution you've posted is something we've considered, and i am assuming will work, but it is not generic and we would not be able to use the `partOf` approach in a generic way. Also, i am assuming that `Documents` and `Pages` should not use `partOf`, because if `Page` uses `partOf` then any change to `Document` triggers the same warning. Going away from generic `partOf` is probably not relevant for the example i've posted, but in our domain we have a need for such an abstraction.

As far as i'm gathering from the conversation with regards to the initial question - the solution then would be to favor more trips to the server (two step process from above) instead of having deep indexing of related data?

**Reply by ayende** [reply]
Yes, if those are the choices, another query is cheaper
